process_scheduling_algorithms/preemptive_priority_scheduling.c

```diff
@@ -0,0 +1,112 @@
+#include <stdio.h>
+
+// Process structure containing ID, arrival time, burst time, priority, executed time, and completion time
+struct Process {
+    int id;
+    int arrival_time;
+    int burst_time;
+    int priority;
+    int executed_time;
+    int completion_time;
+};
+
+// Function to perform preemptive priority scheduling
+void preemptivePriority(struct Process processes[], int n) {
+    int currentTime = 0;
+    int executedProcesses = 0;
+    int currentProcess = -1;
+    float totalTurnaroundTime = 0;
+    float totalWaitingTime = 0;
+
+    while (executedProcesses < n) {
+        int highestPriority = 9999; // A large number as initial priority
+        int nextProcess = -1;
+
+        for (int i = 0; i < n; i++) {
+            if (processes[i].arrival_time <= currentTime && processes[i].executed_time < processes[i].burst_time) {
+                if (processes[i].priority < highestPriority) {
+                    highestPriority = processes[i].priority;
+                    nextProcess = i;
+                }
+            }
+        }
+
+        if (nextProcess != -1) {
+            if (currentProcess != -1 && currentProcess != nextProcess) {
+                printf("Time %d: Process %d -> Process %d\n", currentTime, currentProcess + 1, nextProcess + 1);
+            }
+            currentProcess = nextProcess;
+            processes[currentProcess].executed_time++;
+            currentTime++;
+
+            if (processes[currentProcess].executed_time == processes[currentProcess].burst_time) {
+                executedProcesses++;
+                processes[currentProcess].completion_time = currentTime;
+                printf("Time %d: Process %d completed\n", currentTime, currentProcess + 1);
+                totalTurnaroundTime += (float)processes[currentProcess].completion_time - processes[currentProcess].arrival_time;
+                totalWaitingTime += (float)processes[currentProcess].completion_time - processes[currentProcess].arrival_time - processes[currentProcess].burst_time;
+                currentProcess = -1;
+            }
+        } else {
+            currentTime++;
+        }
+    }
+
+    float avgTurnaroundTime = totalTurnaroundTime / n;
+    float avgWaitingTime = totalWaitingTime / n;
+
+    printf("\nAverage Turnaround Time: %.2f\n", avgTurnaroundTime);
+    printf("Average Waiting Time: %.2f\n", avgWaitingTime);
+}
+int main() {
+    struct Process testCase1[] = {
```

Comment:
Test cases should be seperated to another function named
```c
static void test() {
   ...
}



```
And should be called in the main function

Reply:
@realstealthninja  Thank you for your review I have made the changes suggested by you.